subiquity/models/subiquity.py

```diff
@@ -278,6 +280,10 @@ def _cloud_init_config(self):
             if getattr(model, 'make_cloudconfig', None):
                 merge_config(config, model.make_cloudconfig())
         merge_cloud_init_config(config, self.userdata)
+        # Raises cloudinit.config.schema.SchemaValidationError or
+        # SchemaDeprecationError to alert to merged-cloud-config issues
+        # in installer environment instead of after installed target reboot.
+        validate_cloudconfig_schema(config, strict=True)
```

Comment:
This spot is good, because it will catch the mix of what Subiquity generates and what the user provides, if anything.

It's a bit late in the install process though - there is a chance that someone has provided
```
autoinstall:
    user-data:
        something: invalid
```
And then then don't hear about it until things are almost done.

What's your thoughts about an earlier check for this, in subiquity/server/controllers/userdata.py ?

Reply:
@dbungert How will a warning log be visible to the user if generated up in subiquity/server/controllers/userdata.py ? I'm reading this as a warning that happens to be emitted only to /var/log/installer/* files and isn't really "seen" in UX. Is that incorrect?

Reply:
We can move it earlier in the install process and we probably don't want to specifically render the full details of the schema failure because that then results in leaked sensitive content into /var/log/installler/* which may be visible on the target system to non-root users I think. The changeset I just pushed basically gives us a redacted warning saying: `Invalid cloud-config provided: Please run 'sudo cloud-init schema --system' for details`. This message would be applicable in the installed system after reboot, but not in the installer/ephemeral environment. Maybe we want to try/except SchemaValidationError in ` subiquity/server/controllers/userdata.py` with a concrete warning about the provided invalid `autoinstall.user-data` key path (and redact the offending value) to point folks specifically to the part of the autoinstall that looks like it will be a problem for cloud-init.

Reply:
@dbungert here's a diff of what I was thinking. The problem a live-installer environement has, I think, is that /autoinstall.yaml is  renderered with any cloud-init passthrough user-data  in a top-level `user-data` key right?

I'd like to be able to give a constructive warning message for the ephemeral environment additionally saying something like run `sudo cloud-init schema --config-file /autoinstall.yaml` but cloud-init will not validate the top-level `user-data` key as valid cloud-config.

So, we'd need a YAML artifact emitted that is only the content below autoinstall.user-data. Maybe we emit this on schema error we could emit some file `/autoinstall-invalid-userdata.yaml` which the warning message could point to.


If the folloowing invalid autoinstall is provided:
cat > brokenautoinstall.yaml <<EOF
#cloud-config
autoinstall:
   version:1
   user-data:
       chef: 
           install_type: 1
EOF

Then a `sudo cloud-init schema -c /autoinstall-invalid-userdata.yaml` will be able to emit errors like:
```
$ sudo cloud-init schema -c /autoinstall-invalid-userdata.yaml
Error:
Cloud config schema errors: chef.install_type: 1 is not of type 'string', chef.install_type: 1 is not one of ['packages', 'gems', 'omnibus']
```

The above invalid autoinstall.yaml would generate the following WARNING with the diff below:
WARNING: Invalid autoinstall config schema provided for keys:autoinstall.user-data.chef.install_type"

But there is no easy way currently to tease out the autoinstall.user-data for cloud-init schema validation on the commandline so we can't provide helpful breadcrumbs to lead someone to discover specifically what the schema prob is.

diff --git a/subiquity/server/controllers/userdata.py b/subiquity/server/controllers/userdata.py
index 77e743e0..efa76f52 100644
--- a/subiquity/server/controllers/userdata.py
+++ b/subiquity/server/controllers/userdata.py
@@ -13,9 +13,22 @@
 # You should have received a copy of the GNU Affero General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+import logging
+
+from cloudinit.config.schema import (
+    SchemaValidationError, validate_cloudconfig_schema
+)
+
 from subiquity.server.controller import NonInteractiveController
 
 
+log = logging.getLogger("subiquity.server.controllers.userdata")
+
+INVALID_USERDATA_SCHEMA_TMPL = (
+   "Invalid autoinstall config schema provided for keys: {invalid_keys}"
+)
+
+
 class UserdataController(NonInteractiveController):
 
     model_name = 'userdata'
@@ -26,6 +39,19 @@ class UserdataController(NonInteractiveController):
         }
 
     def load_autoinstall_data(self, data):
+        # We are processing autoinstall.user-data valid raw cloud-config to
+        # give a sensible warning to the user based on content they provided.
+        try:
+            validate_cloudconfig_schema(data, strict=True)
+        except SchemaValidationError as exc:
+            invalid_keys = [
+                f"autoinstall.user-data.{schema_path}"
+                for (schema_path, _msg) in exc.schema_errors
+            ]
+            message = INVALID_USERDATA_SCHEMA_TMPL.format(
+                invalid_keys=", ".join(invalid_keys)
+            )
+            log.warning(message)
         self.model.clear()
         self.model.update(data)